src/com/amazon/ion/impl/IonReaderNonContinuableSystem.java

```diff
@@ -0,0 +1,352 @@
+        if (ivmSid != null) {
+            // An IVM has been found between values.
+            if (typeAfterIvm == null) {
+                // Only save the type of the next user value the first time an IVM is encountered before that value.
+                typeAfterIvm = type;
+            }
+            // For consistency with the legacy implementation, the system reader surfaces IVMs as symbol values.
+            type = IonType.SYMBOL;
+            pendingIvmSid = ivmSid;
+            return true;
+        } else if (pendingIvmSid != -1) {
+            // All preceding IVMs have been surfaced. Restore the value that follows.
+            pendingIvmSid = -1;
+            type = typeAfterIvm;
+            typeAfterIvm = null;
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public IonType next() {
+        if (handleIvm()) {
+            // This happens if there were multiple IVMs between top-level values. They will be drained from the queue
+            // one-by-one on each call to next().
+            return type;
+        }
+        if (reader.nextValue() == NEEDS_DATA) {
+            if (handleIvm()) {
+                // This happens if one or more IVMs occurs before the end of the stream.
+                return type;
+            }
+            reader.endStream();
+            type = null;
+        } else {
+            type = reader.getType();
+            handleIvm(); // Handles the case where one or more IVMs occurred before the value.
+        }
+        return type;
+    }
+
+    @Override
+    public void stepIn() {
+        reader.stepIntoContainer();
+        type = null;
+    }
+
+    @Override
+    public void stepOut() {
+        reader.stepOutOfContainer();
+        type = null;
+    }
+
+    @Override
+    public int getDepth() {
+        return reader.getDepth();
+    }
+
+    @Override
+    public IonType getType() {
+        return type;
+    }
+
+    /**
+     * Prepares a scalar value to be parsed by ensuring it is present in the buffer.
+     */
+    protected void prepareScalar() {
+        IonCursor.Event event = reader.getCurrentEvent();
+        if (event == IonCursor.Event.VALUE_READY) {
+            return;
+        }
+        if (event != IonCursor.Event.START_SCALAR) {
+            // Note: existing tests expect IllegalStateException in this case.
+            throw new IllegalStateException("Reader is not positioned on a scalar value.");
+        }
+        if (reader.fillValue() != IonCursor.Event.VALUE_READY) {
+            throw new IonException("Unexpected EOF.");
+        }
+    }
+
+    @Override
+    public IntegerSize getIntegerSize() {
+        if (getType() != IonType.INT) {
+            return null;
+        }
+        prepareScalar();
+        return reader.getIntegerSize();
+    }
+
+    @Override
+    public boolean isNullValue() {
+        return pendingIvmSid == -1 && reader.isNullValue();
+    }
+
+    @Override
+    public boolean isInStruct() {
```

Comment:
Did you consider inheriting from the `IonReadeContinuableCore` for the below? I'm sure there's downside there, just curious if you looked at it.

Reply:
Yeah, I wanted to delegate to IonReaderContinuableCore (an interface), rather than inherit from the  IonReaderContinuableCoreBinary implementation because that allows us to use this same system wrapper to wrap either IonReaderContinuableCoreBinary or a potential future IonReaderContinuableCoreText.